File: neoforge/src/main/java/com/stal111/forbidden_arcanus/data/DataGenerators.java
package com.stal111.forbidden_arcanus.data;

import com.stal111.forbidden_arcanus.ForbiddenArcanus;
import net.minecraft.data.DataGenerator;
import net.minecraft.data.metadata.PackMetadataGenerator;
import net.minecraft.network.chat.Component;
import net.minecraft.world.flag.FeatureFlagSet;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.data.event.GatherDataEvent;

/**
 * @author stal111
 * @since 2021-01-26
 */
@EventBusSubscriber
public class DataGenerators {

    @SubscribeEvent
    public static void gatherData(GatherDataEvent.Client event) {
        DataGenerator generator = event.getGenerator();

        DataGenerator.PackGenerator featurePack = generator.getBuiltinDatapack(true, ForbiddenArcanus.MOD_ID, "preview");

        featurePack.addProvider(output1 -> PackMetadataGenerator.forFeaturePack(
                output1,
                Component.literal("Enable experimental features for Forbidden Arcanus"),
                FeatureFlagSet.of(ForbiddenArcanus.PREVIEW)
        ));
    }
}

File: datagen/src/main/kotlin/com/stal111/forbidden_arcanus/datagen/DataGenerators.kt
package com.stal111.forbidden_arcanus.datagen

import com.stal111.forbidden_arcanus.ForbiddenArcanus
import net.neoforged.bus.api.IEventBus
import net.neoforged.fml.common.Mod
import net.valhelsia.dataforge.DataForge

@Mod(ForbiddenArcanus.MOD_ID)
class DataGenerators(eventBus: IEventBus) {

    init {
        DataForge.setup(ProviderCollector(), eventBus)
    }
}